### DISCUSSION TITLE
0.15 Retrospective

### DISCUSSION THREAD

**Discussion by Trashtalk217** [bevyengine/bevy#16431] [General]
The release cycle is quickly nearing it's end. During this cycle, some things went well, while others went poorly. Leave your feedback here, so we can hopefully improve the process for upcoming release cycles!

**Comment by Trashtalk217** [comment]
Writing release notes / migration guides is currently a huge slog. Especially for @alice-i-cecile and @cart since they have to write the release notes for the feature authors that don't do it themselves (which is quite a lot). Maybe it's an idea to include the writing of release notes before merging PRs? Make it as standard as writing documentation and tests?

**Reply by bushrat011899** [reply]
This is ties in with our examples. Ideally, features are demonstrable, and currently that's either as a snippet in the PR body (have to copy-paste to run locally), or a new/modified example in the repo. Not every feature should have a whole example just to demonstrate it. Ideally, we could put these snippets in a release note document where it can be tested and documented. Once that feature releases, that code wont be compile-checked any more, and instead be kept for historical documentation.

**Reply by alice-i-cecile** [reply]
Moving the release notes / migration guide content into the repo itself (as seperate files, like it currently is) would help make the tracking / coordination / overhead much easier. I've opened https://github.com/bevyengine/bevy/discussions/16430 to discuss strategies for that.

I think that requiring release notes (draft, not final) and migration guides before merging is the way. The current process leaves Cart and I with a huge backlog of work to grind through while all other progress is stalled.

**Reply by aevyrie** [reply]
There have been a few features where I would've *preferred* if I was asked to write a draft release section as part of the PR.
- It's fresh in my mind
- I'm already motivated and working on it
- It helps point out flaws in the public API if I have to write a short snippet for the release, and I'm not happy with it

I don't think this has to be a huge amount of work, but I'd also understand if some contributors were not comfortable writing that sort of content.

**Reply by alice-i-cecile** [reply]
Yeah, and I'm fine to help pitch in for the contributors who aren't comfortable with that. I'd much rather write it fresh, without needing a whole separate PR anyways. My motivation here is 50% encouraging contributors to help more, and 50% reducing delay and process pain.

**Reply by bushrat011899** [reply]
I think it's also important to consider that the in-PR version of release notes and migration guide would be a _first draft_, with _heaps_ of time to iterate in separate follow-up PRs. While it would be ideal for every PR to have a good set of release notes, even a basic "I added `Foo` so you can `Bar` now." is better than nothing.

**Reply by NthTensor** [reply]
I would much rather write the docs/migration-guide/release-notes while it's fresh in my mind. I also know I absolutely won't manage to do that consistently unless I'm required to. The current level of friction (having to open a second PR in a second repo) doesn't help.

**Reply by cart** [reply]
I also like the idea of frontloading this work. Currently I'm liking the idea of putting release note markdown somewhere in the Bevy repo and requiring it as part of the review process. The trick is that we can't / shouldn't enforce this everywhere, so how do we build the process in such a way that (1) we don't encumber non-feature changes / changes that don't need release notes and (2) we don't accidentally _drop_ features by not forcing them through the process.

**Reply by bushrat011899** [reply]
Perhaps the `M-Needs-Release-Notes` (and friends) labels could trigger the inclusion of a GitHub Action which checks if a markdown document has been created with the particular PR number? Based on this [StackOverflow](https://stackoverflow.com/questions/62325286/run-github-actions-when-pull-requests-have-a-specific-label) response, it certainly seems possible to have CI conditionally enforce its inclusion.

**Reply by mockersf** [reply]
There's a CI action that checks that PRs with the label `M-Needs-Migration-Guide` have the proper section in their description

In the same way, we could check on a label if the release note markdown file is present on label `M-Needs-Release-Note`

I think it's OK to require the file to be present, but a PR shouldn't be blocked on its content being final

**Reply by bushrat011899** [reply]
> I think it's OK to require the file to be present, but a PR shouldn't be blocked on its content being final

Agreed. I think a first-draft is sufficient to improve the status quo massively. Suggested changes and follow-up PRs can refine these documents after the fact anyway.

**Comment by Trashtalk217** [comment]
It's really hard to exhaustively / automatically test rendering features across platforms. Are there any CI solutions that could be implemented. Require more extensive tests from feature authors?

**Reply by alice-i-cecile** [reply]
[Pixel Eagle](https://pixel-eagle.com/project/B04F67C0-C054-4A6F-92EC-F599FEC2FD1D) by @mockersf is the start of a cross-platform screenshot based tool to better test rendering.

I'd also like to build out better tools to test rendering features within a single binary: recompiling for each feature sucks. The [`testbed` examples](https://github.com/bevyengine/bevy/tree/main/examples/testbed) are a good start towards that.

Simply asking for more tests from feature authors won't work: we need a real framework to write good rendering test code in rather than just saying "figure it out" on an ad hoc basis.

Relatedly, folks are itching to drop webgl2 support, which will *help* with random breakage (it's very limited), but won't fully resolve things. Rendering will continue to be hard to test (manually and automatically) and prone to platform-specific breakage, and we need to design solutions with that in mind.

**Reply by clarfonthey** [reply]
I do wonder if it'd be possible to support some kind of headless rendering tests. Even if it can't check driver support, it should be able to at least write to a texture and verify it's output, to make sure that at least rendering isn't catastrophically broken.

**Reply by mockersf** [reply]
The issue in tests isn't headless, it's easy enough to simulate a window (just works on macOS and Windows, and a virtual frame buffer on Linux) or indeed just run headless.

The difficulty is having either a GPU usable in tests, or CPU rendering.
- Metal: macOS hosts in GitHub CI have a virtual GPU ("Apple Paravirtual device") available that just works
- DX12: Windows has a CPU renderer available ("Microsoft Basic Render Driver") but that has very small limits, probably due to the available memory/cpu in GitHub CI
- Vulkan: it's easy to install a software renderer ("llvmpipe") and it works

Our tests in CI already run a few rendering tests, save the result to a screenshot and check it didn't change.

Coverage is hard to extend because:
* Number of features tested: software rendering is slow (the same example that runs in 7 seconds on my local computer runs in 24 seconds on the macOS runner, 3 minutes on the Linux runner and 4.5 minutes on the Windows runner)
* Number of platforms tested: macOS / Metal and Linux / Vulkan are easy to test. wasm / WebGL2 and wasm / WebGPU are a bit harder but also possible. Windows / DX12 is broken, and mobile is more complicated (and also broken at the moment, at least in CI)

A solution for those difficulties could be to throw money at it, either by hosting dedicated runners somewhere with the hardware we need or by paying for larger runners on GitHub (it should work, but I haven't actually tested that)

The current tests executed are:
* Every 6 hours (if there are commits), every example is executed in CI on linux and macos, and the result can be seen there: https://thebevyflock.github.io/bevy-example-runner/index.html. Some examples are not deterministic and the screenshot will change at every execution. This needs to be manually checked
* Every push to a PR, the testbed examples are run in CI on macOS, and the screenshot are compared here: https://pixel-eagle.com/project/B04F67C0-C054-4A6F-92EC-F599FEC2FD1D. The testbed examples are made to be fully deterministic and can be relied on for automatic comparison
* Every attempt to merge a PR, the testbed examples are run in CI on Linux and Windows, and the screenshot compared (same link as above). Again, fully deterministic, but only on Linux, not on Windows

The last two points currently are still checked manually because they were added recently and I wanted to check their reliability before adding reporting to them. They've been working well for a few weeks, so it's probably time to take them to the next step, which I'll document soon in https://github.com/bevyengine/bevy/issues/15918

**Reply by djeedai** [reply]
There's one major blocker for me to writing (unit/feature) tests, as opposed to end-to-end ones, and this is winit and the inability to use the default test harness due to tests not running in the main thread. This prevents testing small pieces of GPU code which run a lot faster (few seconds) than a full featured end-to-end test, even on software renderers. This has been and continue to be a major pain point on Hanabi, and I have to rely to visual testing of examples, which is orders of magnitude slower and more tiring. I do have limited GPU tests under a feature flag, which test things not tied to the render world. But that's all I can do for now. And that leaves a big gap in coverage (I can't get past the 50% coverage point, and that's not due to lack of time).

One word on visual testing via screenshots / image comparison: from (professional) experience, the time investment to setup and maintain is not worth the marginal coverage increase compared to targeted manual testing on each change, because you're testing Bevy as a whole when you do that, and any breakage in other parts can break the end result on screen without the render code actually being the issue. Writing actual GPU tests (write data to GPU, execute GPU commands on data, download resulting texture or buffer content back to CPU, and validate data) provides a lot more bang for bucks.

**Comment by Trashtalk217** [comment]
The retained rendering PR turned out to be a bit of a problem. Merged near the end of the cycle, it caused quite a lot of bugs. These bugs had to be ironed out on main which is extremely frustrating for everyone involved.

Since I was pretty heavily involved, I'll give my perspective first, including what I could've done better / how the process could be improved.

## The Initial Problem

In 0.14, at the end of every frame, the render world purged all its entities. For complicated ECS reasons which I'm not going to explain (you can get a jist [here](https://github.com/bevyengine/bevy/issues/12144)), this blocks relations. Because I (and a lot of other ecs folks) care about that, I set out to try and fix it.

## The Process

I made my [first attempt](https://github.com/bevyengine/bevy/pull/14252) at July 9. The exact way I did it was not really important, because I barely got anywhere before I managed to nerdsnipe @re0312, who created [this PR](https://github.com/bevyengine/bevy/pull/14449) on July 23. I want to take one sentence to say that @re0312 is a much better coder than I am and if it weren't for them I likely wouldn't have gotten anywhere.

This PR went through the review wringer, and while I quite often asked around in the Discord for reviews I was often told that people were busy (which is completely fair, large parts of this were during the summer vacation, so I understand that people had better things to do). People are busy, that's just how it is.

The PR became quiet after a bit, so on September 20 I decided to push it [the final stretch](https://github.com/bevyengine/bevy/pull/15320). I figured it'd be a shame if all that work went stale. And I should add that this PR grew stale very quickly because it touches almost every single part of the rendering machinery. In any case, I tried to hurry the merging of this PR, in large part because I was done with it. This included a tactic where I tried to only get a few important reviews from SMEs / Alice, because more input would again slow the PR down. When it eventually did merge, I figured that there were probably still bugs in the code. Places I or re0132 forgot, although I hoped that everything would go off without a hitch. Or at least, I hoped that the bugs wouldn't be *that bad*.

Hope is not a great code review strategy.

In any case, I wrote up a [follow-up issue](https://github.com/bevyengine/bevy/issues/15459) and largely pulled my hands from it, because I was very busy and my attention was needed elsewhere.

## What Went Wrong

There turned out to be a bunch of bugs. Some of them I had foreseen (entity leaks, wonky extraction), some of them I had theorized but deemed unlikely (removing only a synced component from an entity instead of removing the entity leaves said entity behind in the render world). By and large I don't think there was much that was technically insurmountable given the design, but it sure was annoying to track down every single bug.

I'm very sorry to and thankfull for the people that did.

On a personal level, I got inpatient, and tried to force through a PR that clearly wasn't ready. On a slightly deeper level, I didn't know how I could ever test the fullness of the change without just pushing it to main and having a thousand eyes create issues. This PR clearly needed a strategy for transitioning to a retained render world, but I don't know how that would've looked like. Please let me know, if you do.

I think that this PR has kind of flew under the radar (in the end that was kind of intentional). This change is important for the ECS folks, but largely involves rendering code. They Rendering people seemed less interested in this refactor. The fact that I took this on without having every played with the rendering internals before, should probably be noted somewhere. Also, because this was a refactor, I think a lot of people who looked at it, underestimated how much could break, right untill everything broke.

I'm out of thoughts for now, if anyone has questions, let me know.

**Reply by NthTensor** [reply]
For what it's worth, I don't think the issue with this PR was impatience, and I think it was really useful to have a champion to push it through. It was not a mistake to merge this. Yes, it would have been better to land it early in 0.16, but I think there was a good chance it would have gone stale if we had tried to hold it off, maybe burned out the authors, and that would have been worse.

Ideally we should have held off on merging it until after the release (maybe a month) and blocked all other non-essential rendering changes until after it was merged, to minimize churn.

IMO, that's the real issue. What do we do with PRs that are (a) through the review process and "ready-to-merge" but (b) can't/shouldn't be merged now and (c) will quickly go stale? Is there a way we can put them "on ice" and minimize merge conflicts for a while?

**Reply by Trashtalk217** [reply]
It's an interesting idea, but blocking most rendering PRs untill one specific refactor is finished sounds like it's gonna become an annoyance rather quickly. For exceptional cases, it could maybe be done. In practice this would mean that Alice would have to hold on pressing the button on any rendering PRs. Possible: yes, a bit weird: also yes.

I think the underlying reason why these sort of PRs got finished at the end of a release cycle is because deadlines are a thing that humans respond to. I wouldn't have pushed as much as I did if I knew there was still an entire release cycle left to go. The original plan was to get this merged at the start of 0.15, and back than I didn't (it also wasn't done then, but still).

**Reply by cart** [reply]
Thanks for the detailed postmortem!

From my perspective the only "failure" here was the failure of our system as a whole to put the breaks on a major architectural change that wasn't yet fully conceptually formed. People with the knowledge required to vet the architecture were not looped in (or at least, not to the degree that they should have been), and we merged it "hoping for the best" because it "had enough reviews". For core architectural changes, "having enough reviews" is not enough. It _must_ be about "having the _right_ reviews".

>  It was not a mistake to merge this.

I strongly disagree. In this case, many of us knew the "statefulness" of the retained render world had a lot of hard problems (and many of the "discovered" issues were issues that had already been enumerated). The reason we've been putting it off for so long is that "Retained Render World" is a hard problem. By merging in its current state, we committed to a hard road without knowing exactly where it lead or how we planned to walk it. Even after all of the bug fixing we are (still) standing on flimsy ground as many of the bugfixes are one-off per-feature hacks, rather than holistic solutions to general problems. Until we are back on solid ground we will continue to accumulate tech debt and introduce new regressions because we've made "airtightness" a manual process. We've written a check that we can't yet cash and we're paying interest. Waiting until we actually built a solid platform to stand on (or at least knew what it looks like), was absolutely the right move in this case.

I should have stepped in, provided relevant information, and pumped the breaks until we had a finished architecture. I was busy with other things and this slipped in. I will increasingly be busy with other things / can't be everywhere, so the question is, how do we identify cases like this going forward and ensure somebody applies the breaks for major architectural changes that haven't been approved by the relevant people.

As always, there is no value in pointing fingers here. We should be discussing processes. In this case, it seems like the broken process was that a highly controversial / core piece of architecture received the `X-Blessed` tag (and was merged that way), despite not having two rendering SME approvals (or alternatively, my approval).

**Reply by cart** [reply]
I'll also say that I'm _very_ happy that @Trashtalk217 chose to work on (and champion) this effort. I do think retained is the right path to be on. I just personally wish that our process forced this to wait until it was fully baked.

**Reply by NthTensor** [reply]
> By merging in its current state, we committed to a hard road without knowing exactly where it lead or how we planned to walk it.

I agree that, in the set of all possible worlds, the one where we wait and this merges smoothly and stably without breakages is the best. When I say "it was not a mistake to merge this" it's coming from concern that we would never have in fact merged if we waited, and that a long architectural review process would have ground it to a halt.

We took a calculated risk, it didn't pay off. In my mind, and I'm sure in the minds of the users eager to reap the new benefits it unblocks in the form of cold-specialization and relations, it was worth it. I was surprised that it merged, but I will continue to defend it's merging on the general principle that it is sometimes OK to break eggs when making an omelet, especially when the chefs are working for free and would otherwise be left tending the egg for potentially months without end. Otherwise you may find yourself running out of chefs.

All that said, it does seem like there was a miscommunication in the process. Do I understand that the pr was tagged "Blessed" because the overall idea of retained rendering had been accepted as a goal, while the specific implementation had still not been fully vetted? If so, that does seem like a mistake.

**Reply by mockersf** [reply]
"blessed" is supposed to mean the controversial nature of something has been resolved, not that it's on a fast track to being merged. it should still follow the nomal process to be merged, with enough reviews by people with enough expertises

we could have had three source of controversy on that change:

1.  do we want to do it or not?
2. do we want to do it that specific way?
3. is it ready and covers everything?

1 was blessed, I haven't followed the discussion enough to know about 2, and 3 could have been resolved with more reviews without going through being controversial

**Reply by alice-i-cecile** [reply]
Yeah, there was agreement in principle from both ECS and rendering leadership, but the changes and migration path and implications were not adequately reviewed and discussed.

**Reply by cart** [reply]
> It's coming from concern that we would never have in fact merged if we waited, and that a long architectural review process would have ground it to a halt.
> We took a calculated risk, it didn't pay off

I do understand these concerns, and acknowledge that "efforts dying out via leadership neglect" has been and continues to be a problem to solve (although I _hope_ we can all agree that things have gotten better recently, which is a result of "letting things happen" that we would not have in the past). We were _very_ lax last cycle when it came to reviews. In many ways it was good as it let people "just get things done" / let the project grow. But this is me sounding the alarm bells on something that I strongly believe went too far. If we allow architectural choices in this category to be made in this way, the quality and vision of the project will de-cohere in a way that makes me deeply uncomfortable. I would rather we do not move forward at all (for core architecture things), and yes, risk burning contributors out, than move forward in this way. If the SMEs are not steering the ship, it should not move. If the ship isn't moving fast enough, we either need to strategically and intentionally stop while SMEs are off doing other things, or we need to find more SMEs.

**Reply by cart** [reply]
@NthTensor I do also want to call out that I appreciate that people have pushed back against me being "too precious" with the code in the past (and this includes you!). I was standing in the way of progress. And I'm glad there are still people applying counter forces and challenging my assumptions. But I still strongly believe I am right about cases like this, and I intend to hold this line going forward. Bevy is good in a large part because we are careful and thoughtful about architecture, and because we have held lines like this in the past.

**Reply by NthTensor** [reply]
Absolutely! We have drastically improved the contribution process over the last year, and I am deeply appreciative of the choices that project leadership made to achieve that increase in velocity. We clearly benefit from your oversight, and the tools at your disposal to prevent decoherence and destabilization must be kept in good working order. Stability/velocity is a delicate balance, but it's clear from these responses that our position on that scale is something the leadership treats with great care and consideration. So long as that's true where I would draw the line in this particular case is really beside the point, and I for one am satisfied.

**Comment by alice-i-cecile** [comment]
I'm really pleased about the success of working groups. Continuing the trend started with bevy_color, the text rework, bevy_picking, curves and the next generation scene working groups have all been really effective at delivering high quality work in a cohesive and collaborative way.

Having explicit blessing, a design, and a place to talk really helped!

**Reply by BD103** [reply]
I agree with this! I've had a great experience joining working groups and helping work on larger goals. Working groups are great to find contributors with similar interests and collaborate with them on issues in a semi-official manner.

**Reply by occuros** [reply]
Is there currently a way to know which topics are considered for a working group?

It's easy to discover the ones which have already been formed, but not sure how to find potential future ones.

**Reply by NthTensor** [reply]
I kinda don't think we should have a list? Much prefer that they come about organically because people have been thinking about/struggling with something in the engine. Anyone can propose a group, and any proposed group will get consideration for scope & alignment by the leadership team.

Agree btw, I continue to be really happy with the wg process. I think we are still struggling to get people to write down & document their designs, but maybe that will come with changes to release notes.

**Reply by Trashtalk217** [reply]
I want to give a little bit of pushback with regards to the relations working group. As far as I can tell we do have a clear goal before us (both an rfc and a hackmd document), but the work has been somewhat slow. I mostly blame this on a lack of manpower. James has been busy at work, same goes for me and periwink.

ECS threads also have a rather unique ability to derail into hypotheticals like: 'could you build a relational database with relations?' or 'How should we handle *thing that flecs hasn't even implemted yet and is still at least a year away*?' I understand that I can't stop people from talking about things they're passionate about, but it does kind of annoy me sometimes. Still, the relations working group is no worse than the ecs-dev channel, so it's a wash on that point.

I'm kind of wondering if the 'self-contained task' this working group is doing is too large, but I'm not sure. From where I'm standing, landing an MVP for relations is possible in 0.16, but I said the same about 0.15, so I might just be delusional.

**Reply by brandon-reinhart** [reply]
On the outside, the working group threads were helpful as an early adopter. I had a place I could ask questions related to an implementation detail or bug or give feedback on the feature, etc.

**Comment by alice-i-cecile** [comment]
Upstreaming bevy_picking went really well overall, and I think it's a model to emulate. This is a really effective way to rapidly expand our capabilities, and it helps share the maintenance burden for vital functionality.

I think in the future we should consider polishing / improving further in the upstream repo (docs and examples especially), and I underestimated the difficulty of the work: it ended up taking fairly experienced engineers to do the integration work effectively.

**Reply by alice-i-cecile** [reply]
@NthTensor I'd be interested in your thoughts on this briefly, as you ended up largely leading this effort. @Jondolf and @aevyrie as well, since you were also major contributors / the upstreamed crate's author.

**Reply by aevyrie** [reply]
I thought it went really well. I was away for personal reasons for the majority of the process, but it went smoothly from my perspective.

I get the sense this area of bevy could use a SME; there is a lot of surprising complexity, and even after 0.15 there is still a good amount before we have "finished". It's a bit difficult to feel like the decisions you make and the design you are settling on is going to stick without a dedicated SME present. My biggest worry is that without someone with depth of knowledge, context, and a clear vision, this code will begin to rot from well-meaning but poorly thought out changes.

**Reply by NthTensor** [reply]
Agree with @aevyrie here about an SME, although I think Alice serves this roll pretty well. 

I'd like to see us upstream more stuff. It's a challenge, but a different kind of challenge than greenfield work, which mostly comes from solving problems in the ecosystem crate by modifying the ecs or the other core crates (the for bevy_picking this was adding bubbling to observers). IMO it's something you're usually going to need engine-devs for (which was the issue with the student group, not their fault).

For future up streams, this is probably how I'd work it:
1. **Design** and implement necessary engine features, and in parallel add polish to the ecosystem crate.
2. **Copy** all the important files from ecosystem crate verbatim into the engine repo (leave stuff out of the workspace/example tree).
3. **Mock** the up-streamed crate to get it compiling (even if it's broken) and write specs/tests.
4. **Integrate** new ecs features and fix broken stuff until it meets the specs/tests.
5. **Iterate** on the up-streamed api with feedback from early adopters (this was crucial for getting `bevy_picking` ready for release).

I probably wouldn't try to plan more than one of these stages at a time. It's better to bite off manageable chunks.

**Comment by alice-i-cecile** [comment]
The new workflow for writing migration guides / PRs with each piece of content in its own file is great (other than being done later + in a different repo). Almost no merge conflicts, super easy to re-organize, and the new PR merging strategy is fantastic. Really pleased with how that design and tooling work has paid off.

**Reply by cart** [reply]
Agreed! Biggest downside is it breaks hot-reloading so you can't preview your changes inline (instead you have to stop and start `zola serve`). We might be able to resolve this by doing something different on our end, or this might need to be fixed in upstream zola.

**Reply by rparrett** [reply]
At some point in the 0.14 cycle I hacked around the lack of hot reloading with this masterpiece, dug up from my discord history:
```bash
fswatch -0 -o release-content | xargs -0 -n1 -I{} sh -c 'touch content/test.md && rm content/test.md'
```
But I completely forgot about that for 0.15.

Also from that chat:
> I looked into whether it's possible to get zola to hot reload on changes to stuff in release-content. Seems like no, the list of paths it watches are hardcoded.

But configuring additional paths to watch feels like something that they may be okay with. (just guessing)

It's also a huge pain to have to dig up the corresponding file for each section that needs editing while reading through the rendered document. When sections need to be combined, this gets even more tedious / error-prone. I recall starting to read through the draft a while back, realizing how many sections needed rework like this and just giving up.

**Reply by rparrett** [reply]
Just remembered this and threw out a feature request here: https://github.com/getzola/zola/issues/2741

**Reply by cart** [reply]
Good call. Just added some more context

**Comment by mrchantey** [comment]
As usual I am very happy with the response time and quality for PRs, reviews and dev discord questions, it makes the process fun and gives me confidence to keep contributing 🚀

**Comment by djeedai** [comment]
Some breaking change #16163 went in after the last Release Candidate (`.rc-3`) and before the final Release, causing #16563, which requires a new patch release to fix. This means users couldn't test the change before the release. The issue should not have happened: it was trivially caught twice independently and immediately the second we tried to upgrade. We were lucky this time there was an easy workaround, otherwise we would have needed a `0.15.1` right after `0.15.0` (the next day). Breaking changes without RC defeat the purpose of RC. I don't know if it's an overlook (didn't realize the change was breaking) or on purpose (didn't want to delay further so skipped `-rc.4`) but this was not ideal, and at least in my case caused churn and lost time which were otherwise not used to tackle other issues (as of today Hanabi still didn't release for 0.15; this is not the only reason, but this contributed).

**Reply by cart** [reply]
I'll take responsibility for this one! I wanted this in, didn't think it was breaking in this way, and didn't want to delay the release further. Doing another RC and delaying would have been safer, and we probably should have done that.

The problem is that we didn't wire up the optional `bevy_image` dependency to the top level `bevy` crate. There is a workaround. Just add a manual dependency on `bevy_internal` and set the optional feature there. 

```toml
[dependencies]
bevy = { version = "0.15", default-features = false }
bevy_internal = { version = "0.15", features = [ "bevy_image" ]}
```

**Reply by djeedai** [reply]
Thanks @cart, I went with the workaround of adding a dependency to `png` instead, to not delay the release.

It would be nice to have guidelines around the usage of features. I started a discussion on Discord but there was no real consensus nor even strong opinion on what to do (_e.g._ expose all features to top-level, avoid features, other...?) and that creates inconsistencies.

**Reply by alice-i-cecile** [reply]
Yeah, I'd love to have a section on features in the contributing guide TBH.

**Reply by bushrat011899** [reply]
Agreed regarding feature management. 0.16 is going to include even more features due to the `no_std` work, and I think there's a lot of room for improvement in how we document features, and standards around them.

A particular pet peeve of mine has been feature discovery. The only reliable way to see what features a crate has is to just read the `Cargo.toml`, but I'll still miss that an optional dependency is being used as a feature (I hate this Cargo feature personally but that's another discussion).

**Reply by alice-i-cecile** [reply]
Feature discovery / documentation is a rough spot in Rust in general; let me ask around to see if there's anything in the works there.

**Reply by NthTensor** [reply]
There is a flags page on docs.rs, which is what I usually reference. 

https://docs.rs/crate/bevy/latest/features

**Reply by mockersf** [reply]
For Bevy we build this page: https://docs.rs/bevy/latest/bevy/#cargo-features
* not for the subcrates
* no easy way to know the features/dependencies enabled by a feature

There's the https://crates.io/crates/document-features crate/macro that enable more fancy documentation, but doesn't enforce that features are documented

on documentation, the pain points (for me) are:
* hard to know what a feature enables. even harder with Bevy as all the features on the main crate go through a bevy_internal proxy feature
* hard to know which features are exposed by subcrates and not possible to enable through the main crate. I think it makes sense to have some of those but they are almost impossible to discover
